app/src/main/java/de/westnordost/streetcomplete/screens/main/map/QuestPinsManager.kt

```diff
@@ -126,22 +135,31 @@ class QuestPinsManager(
 
     private fun onNewTilesRect(tilesRect: TilesRect) {
         val bbox = tilesRect.asBoundingBox(TILES_ZOOM)
-        viewLifecycleScope.launch {
-            val quests = withContext(Dispatchers.IO) { visibleQuestsSource.getAllVisible(bbox) }
+        pinsUpdateJob?.cancel()
+        pinsUpdateJob = viewLifecycleScope.launch {
+            val quests = withContext(Dispatchers.IO) {
+                synchronized(visibleQuestsSource) {
+                    if (!coroutineContext.isActive) null
+                    else visibleQuestsSource.getAllVisible(bbox)
+                }
+            } ?: return@launch
```

Comment:
Not sure if I took note of that before, but as far as I know, the Dispatchers.IO has a very limited threadpool and this construct blocks one of them.
Is it possible to put the `withContext` just around the `getAllVisible`?

Reply:
My decision for blocking Dispatchers.IO actually was deliberate, but for the same reason.

I didn't find any clear information, but I assumed `viewLifecycleScope.launch` uses [Dispatchers.Default](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-default.html), which only has one thread for each CPU core, while [Dispatchers.IO](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-i-o.html) has up to 64 threads.

Reply:
I see, ok